.github/workflows/pre-commit.yml

```diff
@@ -0,0 +1,76 @@
+name: Lint Check (pre-commit)
+
+on:
+  # We have to use pull_request_target here as pull_request does not grant
+  # enough permission for reviewdog
+  pull_request_target:
+  push:
+
+jobs:
+  pre-commit-push:
+    name: Pre-Commit check on Push
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'push' }}
+
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: 3.13
+
+        # We wish to run pre-commit on all files instead of the changes
```

Comment:
I'm wonder if this is really what we want?  Aside from the obvious inefficiencies involved in checking all files all the time this will make it potentially difficult to *override* a lint/format setting.  Let me give you a concrete case of this.  In my day-job-project we using clang-tidy on code, but it struggles in any setup we've used to nicely format monadic function chains. So if we had are setup run all files every checkin it would be annoying (note there's likely ways to turn off the changes we don't know about).  Looking for other opinions on this point.

Reply:
Upstream for this CI config is in `exemplar`, so the context is this is meant to be run along side some huge unit test complex.

The main idea is if we have a linting rule (or any enforceable rule) then if some part of the system is broken, the whole system is broken. Linting potentially causes sweeping changes due to reformatting, so its best the reformat is as close to lint error as possible (best in the same commit).

Explicitly, we are trying to avoid this:
- commit 0 feature [x diff breaks lint a.cpp]
- commit 1 feature [✓ diff pass lint b.cpp]
- commit 2 ready to ship [✓ diff pass lint b.cpp]
Here the lint failure in 0 may still not be fixed because no diff later touches a.cpp .

- This is inefficient: Yes this is inefficient, but all the linters are light weight and often doesn't even involve the need to construct an AST. In comparison to the unit test of cross platform + cross compiler matrix/ later the long clang-tidy job (in exemplar/ C++ project context), the cost here is minimal. Arguing lint cost here is in efficient.
- Override linting: This is not the concern of CI, overriding linting should be the concern of configuring the linter. e.g. `codespell:ignore`, `markdownlint-disable-next-line`, `gersemi: ignore`. Linting error should be explicitly suppressed  (best in-line), but not to be ignored. Running lint check only on diff doesn't guarantee error propagation unless we do complicated trickery.  Think this as unit test, you would ideally run the entire unit-test set, and the solution to flaky unit tests are, ideally to fix them first, if it doesn't work then skip, but not to ignore them from CI perspective out-right.

Reply:
Thanks for the long explanation.  And while I agree there are overrides possible many of these involve actual source modifications which are sometimes unpleasant for authors, etc.  And, if I'm not mistaken the standard we're setting here will require the revisiting of those standards on each commit for unchanged source.

> Upstream for this CI config is in exemplar, so the context is this is meant to be run along side some huge unit test complex.

Sure.  I'll note however, that none of this is explained in exemplar at the moment.

> The main idea is if we have a linting rule (or any enforceable rule) then if some part of the system is broken, the whole system is broken.

To me there's a large difference between a broken test and some code that doesn't pass formatting checks.  So maybe that makes these 'less enforceable rules'. 

I guess at a minimum I'd like some others to chime in on this.

Reply:
I agree with @wusatosi that enforcement of linter checks at CI time is the most desirable option and this is what we're doing with every other repository in beman so far. We "push left" by allowing developers to get this feedback with the pre-commit tool, but if they don't use this, there's review dog to suggest the changes needed at PR time. I've worked in several repositories that enforce linters and it's been a positive experience.